src/dos/cdrom.cpp

```diff
@@ -268,3 +93,212 @@ void CDROM_Interface::LagDriveResponse() const
 
 	prev_ticks = PIC_Ticks;
 }
+
+void CDROM_Interface_Physical::CdReaderLoop()
+{
+	std::unique_lock lock(mutex);
+	while (true) {
+		waiter.wait(lock, [&]{ return sectors_remaining > 0 || should_exit; });
+
+		if (should_exit) {
+			return;
+		}
+
+		if (!queue.IsRunning()) {
+			queue.Clear();
+			queue.Start();
+		}
+
+		// Copy these to local variables so that we can release the mutex lock.
+		const uint32_t local_sectors_remaining = sectors_remaining;
+		const uint32_t local_sector_start = current_sector;
+		lock.unlock();
+
+		// Slow call to read from the CDROM drive. Avoid holding any locks
+		const std::vector<int16_t> audio_samples = ReadAudio(local_sector_start, local_sectors_remaining);
+		const size_t num_frames = audio_samples.size() / 2;
+		const size_t sectors_read = audio_samples.size() / SAMPLES_PER_REDBOOK_FRAME;
+
+		// Convert from int16_t to float before enqueuing
+		std::vector<AudioFrame> audio_frames;
+		for (size_t i = 0; i < num_frames; ++i) {
+			const int16_t left = host_to_le16(audio_samples[i * 2]);
+			const int16_t right = host_to_le16(audio_samples[(i * 2) + 1]);
+			audio_frames.emplace_back(left, right);
+		}
+
+		queue.BulkEnqueue(audio_frames, audio_frames.size());
+
+		lock.lock();
+		if (mixer_channel && !is_paused && sectors_remaining > 0 && queue.IsRunning()) {
+			mixer_channel->Enable(true);
+		}
+		if (sectors_remaining == local_sectors_remaining && current_sector == local_sector_start) {
+			sectors_remaining -= sectors_read;
+			current_sector += sectors_read;
+		}
+	}
+}
+
+void CDROM_Interface_Physical::InitAudio()
+{
+	if (mixer_channel) {
+		return;
+	}
+
+	bool found_unique_name = false;
+	std::string name;
+	for (int i = 0; i < 32; ++i) {
+		name = std::string(ChannelName::CdAudio) + "_" + std::to_string(i);
+		if (!MIXER_FindChannel(name.c_str())) {
+			found_unique_name = true;
+			break;
+		}
+	}
+
+	if (!found_unique_name) {
+		LOG_ERR("CDROM_IOCTL: Too many mixer channels");
+		return;
+	}
+
+	auto callback = std::bind(&CDROM_Interface_Physical::CdAudioCallback,
+	                          this,
+	                          std::placeholders::_1);
+	mixer_channel = MIXER_AddChannel(callback,
+	                                 REDBOOK_PCM_FRAMES_PER_SECOND,
+	                                 name.c_str(),
+	                                 {ChannelFeature::Stereo,
+	                                  ChannelFeature::DigitalAudio});
+
+	thread = std::thread(&CDROM_Interface_Physical::CdReaderLoop, this);
```

Comment:
First I'm hearing of this feature:  https://en.cppreference.com/w/cpp/thread/jthread

It looks like a regular thread except it automatically joins on destruction?  There's also something about a `stop_source` which looks like an overly complicated replacement of my `should_quit` boolean :sweat_smile: 

Maybe this is worth using though.  Could you provide an example of how I might use it here?  I'm a bit confused because it says it has an internal `stop_source` but then in their example they're also passing one into the worker function:  https://en.cppreference.com/w/cpp/thread/stop_source

Reply:
Ok, this is the last code comment.  @johnnovak do you have an opinion on this?  I'm just not sure if I understand the benefit to using this feature or how to implement it correctly.

I think maybe it saves a single `thread.join()` call in the destructor, although I believe we still have to stop the `RWQueue` and do other cleanup.

Reply:
I'm not against leaving it like this. From what I understood, jthread is easier on edge-cases like exceptions and other threads stopping, but I'm don't think they are a real issue here. Explicit cleanup might be better here, especially if there are still other things to clean up.

We can always keep it in the back of our minds and/or experiment with them later. This comment originated from me toying around with std::thread and getting the remark from automatic code review that I should use std::jthread instead, because it's safer. But then again, if you use std::thread well, it should not be much of an issue.

Reply:
@weirddan455 Because we need to do some other manual cleanup/synchronisation anyway, I think it's fine as it is.

Thanks for the tip @japsmits, maybe we'll use jthread next time 😄